pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/SnapshotSegmentAbortedTxnProcessorImpl.java

```diff
@@ -0,0 +1,599 @@
+                                    this.topic.getName(), this.sequenceID, positionNeedToDelete);
+                        }
+                        abortTxnSegments.remove(positionNeedToDelete);
+                        //The process will check whether the snapshot segment is null, and update index when recovered.
+                        indexes.remove(positionNeedToDelete);
+                        //Keep index snapshot and update index
+                        updateSnapshotIndex(persistentSnapshotIndexes.getSnapshot(),
+                                indexes.values().stream().toList());
+                    }).exceptionally(e -> {
+                        log.warn("[{}] Failed to delete the snapshot segment, "
+                                        + "whose sequenceId is [{}] and maxReadPosition is [{}]",
+                                this.topic.getName(), this.sequenceID, positionNeedToDelete, e);
+                        return null;
+                    });
+        }
+
+        //Update the indexes with the giving index snapshot and index list in the transactionBufferSnapshotIndexes.
+        private CompletableFuture<Void> updateSnapshotIndex(TransactionBufferSnapshotIndexesMetadata snapshotSegment,
+                                                            List<TransactionBufferSnapshotIndex> indexList) {
+            TransactionBufferSnapshotIndexes snapshotIndexes = new TransactionBufferSnapshotIndexes();
+            return snapshotIndexWriterFuture
+                    .thenCompose((indexesWriter) -> {
+                        snapshotIndexes.setIndexList(indexList);
+                        snapshotIndexes.setSnapshot(snapshotSegment);
+                        return indexesWriter.writeAsync(topic.getName(), snapshotIndexes);
+                    })
+                    .thenRun(() -> {
+                        persistentSnapshotIndexes = snapshotIndexes;
+                        lastSnapshotTimestamps = System.currentTimeMillis();
+                    })
+                    .exceptionally(e -> {
+                        log.error("[{}] Failed to update snapshot segment index", snapshotIndexes.getTopicName(), e);
+                        return null;
+                    });
+        }
+
+        private CompletableFuture<Void> clearSnapshotSegmentAndIndexes() {
+            ArrayList<CompletableFuture<Void>> completableFutures = new ArrayList<>();
+            //Delete all segment
+            while (!abortTxnSegments.isEmpty()) {
+                if (log.isDebugEnabled()) {
+                    log.debug("[{}] Topic transaction buffer clear aborted transactions, maxReadPosition : {}",
+                            topic.getName(), abortTxnSegments.firstKey());
+                }
+                PositionImpl positionNeedToDelete = abortTxnSegments.firstKey();
+                completableFutures.add(persistentWorker.deleteSnapshotSegment(positionNeedToDelete));
```

Comment:
Deleting the `abortTxnSegments.firstKey` is running asynchronously, so there is a problem in the loop block: On the second loop, when we expect to delete the second element, the actual task is to delete the first element

Reply:
Good point out.